File: src/main/java/com/healthpal/api/config/DataInitializer.java
package com.healthpal.api.config;

import com.healthpal.api.model.Role;
import com.healthpal.api.model.User;
import com.healthpal.api.repository.RoleRepository;
import com.healthpal.api.repository.UserRepository;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.CommandLineRunner;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Component;

import java.time.LocalDateTime;

/**
 * DataInitializer - Initialize Database with Default Data
 * 
 * This component runs when the application starts and initializes:
 * - Default roles (ROLE_PATIENT, ROLE_DOCTOR, ROLE_ADMIN, etc.)
 * - Default admin user
 * 
 * This is part of the application startup process
 * 
 * @author HealthPal Team - Person A (Authentication + Admin + Core Architecture)
 * @version 1.0
 */
@Slf4j
@Component
public class DataInitializer implements CommandLineRunner {

    @Autowired
    private RoleRepository roleRepository;

    @Autowired
    private UserRepository userRepository;

    @Autowired
    private PasswordEncoder passwordEncoder;

    /**
     * Run initialization when application starts
     * @param args Command line arguments
     * @throws Exception If initialization fails
     */
    @Override
    public void run(String... args) throws Exception {
        log.info("Starting database initialization...");

        // Initialize default roles
        initializeRoles();

        // Initialize default admin user
        initializeAdminUser();

        log.info("Database initialization completed successfully");
    }

    /**
     * Initialize default roles
     */
    private void initializeRoles() {
        log.info("Initializing default roles...");

        String[] roleNames = {"ROLE_PATIENT", "ROLE_DOCTOR", "ROLE_ADMIN", "ROLE_NGO", "ROLE_DONOR"};

        for (String roleName : roleNames) {
            if (!roleRepository.existsByName(roleName)) {
                Role role = Role.builder()
                        .name(roleName)
                        .description("Role for " + roleName.replace("ROLE_", "").toLowerCase() + " users")
                        .isActive(true)
                        .build();

                roleRepository.save(role);
                log.info("Role created: {}", roleName);
            } else {
                log.debug("Role already exists: {}", roleName);
            }
        }
    }

    /**
     * Initialize default admin user
     */
    private void initializeAdminUser() {
        log.info("Initializing default admin user...");

        String adminUsername = "admin";
        if (!userRepository.existsByUsername(adminUsername)) {
            Role adminRole = roleRepository.findByName("ROLE_ADMIN")
                    .orElseThrow(() -> new RuntimeException("ROLE_ADMIN not found"));

            User adminUser = User.builder()
                    .username(adminUsername)
                    .email("admin@healthpal.com")
                    .password(passwordEncoder.encode("Admin@123"))
                    .firstName("System")
                    .lastName("Administrator")
                    .phoneNumber("+1234567890")
                    .isActive(true)
                    .isEmailVerified(true)
                    .isPhoneVerified(true)
                    .isAdmin(true)
                    .build();

            adminUser.addRole(adminRole);
            userRepository.save(adminUser);
            log.info("Default admin user created: {}", adminUsername);
        } else {
            log.debug("Admin user already exists");
        }
    }
}

File: config/dataInitializer.js
const Role = require('../models/Role');
const User = require('../models/User');
const bcrypt = require('bcryptjs');
const logger = require('../utils/logger');

class DataInitializer {
    async initialize() {
        logger.info('Starting database initialization...');
        try {
            await this.initializeRoles();
            await this.initializeAdminUser();
            logger.info('Database initialization completed successfully');
        } catch (error) {
            logger.error('Database initialization failed:', error);
            throw error;
        }
    }

    async initializeRoles() {
        logger.info('Initializing default roles...');
        const roles = [
            { name: 'ROLE_PATIENT', description: 'Role for patient users' },
            { name: 'ROLE_DOCTOR', description: 'Role for doctor users' },
            { name: 'ROLE_ADMIN', description: 'Role for admin users' },
            { name: 'ROLE_NGO', description: 'Role for NGO users' },
            { name: 'ROLE_DONOR', description: 'Role for donor users' }
        ];

        for (const roleData of roles) {
            const existingRole = await Role.findOne({ name: roleData.name });
            if (!existingRole) {
                await Role.create({
                    ...roleData,
                    isActive: true
                });
                logger.info(`Role created: ${roleData.name}`);
            } else {
                logger.debug(`Role already exists: ${roleData.name}`);
            }
        }
    }

    async initializeAdminUser() {
        logger.info('Initializing default admin user...');
        const adminUsername = 'admin';
        const existingAdmin = await User.findOne({ username: adminUsername });
        
        if (!existingAdmin) {
            const adminRole = await Role.findOne({ name: 'ROLE_ADMIN' });
            if (!adminRole) {
                throw new Error('ROLE_ADMIN role not found');
            }

            const adminUser = await User.create({
                username: adminUsername,
                email: 'admin@healthpal.com',
                password: await bcrypt.hash('Admin@123', 10),
                firstName: 'System',
                lastName: 'Administrator',
                phoneNumber: '+1234567890',
                isActive: true,
                isEmailVerified: true,
                isPhoneVerified: true,
                isAdmin: true,
                roles: [adminRole._id]
            });

            logger.info(`Default admin user created: ${adminUsername}`);
            return adminUser;
        } else {
            logger.debug('Admin user already exists');
            return existingAdmin;
        }
    }
}

module.exports = new DataInitializer();